### ISSUE TITLE
Change in a way to store objects in reducers in frontend

### ISSUE THREAD

**Issue by ChanwooLucioLee** [swsnu/swpp2020-team13#26]
We have stored objects in an array in reducers in frontend.
For example, the initial state of "goals" in goalReducer is [], which is an array.

However, we will start to store objects in an object with its id as a key.
For example, goals = [{'id': 1, 'title': 'TITLE1'}, {'id': 2, 'title': 'TITLE2'}] will be stored as 
goals = {1: {'id': 1, 'title': 'TITLE1'}, {'id': 2, 'title': 'TITLE2'}}

Why is it better?
: You don't need to use filter() to find an object when you edit or delete. goals[<id>] is all you need to do.

How to migrate?
- Backend will still send objects as before. So the response.data (action.payload) will look the same as before.
- We will use lodash library's "mapKeys" function to convert an array to an object with a key as id.
  _.mapKeys(action.payload, 'id'} will do the job. The second argument 'id' indicates that the 'id' value of each item will be the key for the item.
- When you want to convert the object to an array as before, use Object.values(<object>). That will convert 
{1: {'id': 1, 'title': 'TITLE1'}, {'id': 2, 'title': 'TITLE2'}} to [{'id': 1, 'title': 'TITLE1'}, {'id': 2, 'title': 'TITLE2'}].

**Comment by jiyewise** [comment]
I agree with your point that edit and delete would be implemented much easily.

1. In that case, how does add work? In the previous reducer, add was done by adding the new element to the original array by [...state, newItem]. When the states are in a object form, do we have to change it to a list first to add it?

2. If we change the format, I guess then we should change the syntax of not only the reducers but also testing codes, as I have already finished testing on all the actions and reducers. If we change the format before sprint 4 ends, I guess we have to change all the testing codes also.  Would the changes be minimal? Do we have enough time to do this?

**Comment by ChanwooLucioLee** [comment]
1. [...state, [action.payload.id]: action.payload} will do

2. I will not change code for goals and tasks. The new format will only apply to achievements.

**Comment by jiyewise** [comment]
1. Didn't know such syntax also works on objects! Thanks for letting me know 😃
2. I see..! Then I guess there would be no additional time costs. Thank you!
